pageserver/src/tenant/ephemeral_file.rs

```diff
@@ -21,18 +21,42 @@ use crate::tenant::storage_layer::inmemory_layer::vectored_dio_read::File;
 use crate::virtual_file::owned_buffers_io::io_buf_aligned::IoBufAlignedMut;
 use crate::virtual_file::owned_buffers_io::slice::SliceMutExt;
 use crate::virtual_file::owned_buffers_io::write::{Buffer, FlushTaskError};
-use crate::virtual_file::{self, IoBufferMut, VirtualFile, owned_buffers_io};
+use crate::virtual_file::{self, IoBufferMut, TempVirtualFile, VirtualFile, owned_buffers_io};
+
+use self::owned_buffers_io::write::OwnedAsyncWriter;
 
 pub struct EphemeralFile {
     _tenant_shard_id: TenantShardId,
     _timeline_id: TimelineId,
     page_cache_file_id: page_cache::FileId,
     bytes_written: u64,
-    buffered_writer: owned_buffers_io::write::BufferedWriter<IoBufferMut, VirtualFile>,
+    // Always Some except during Drop
+    inner: Option<Inner>,
+}
+
+struct Inner {
+    file: TempVirtualFileCoOwnedByEphemeralFileAndBufferedWriter,
+    buffered_writer: BufferedWriter,
     /// Gate guard is held on as long as we need to do operations in the path (delete on drop)
     _gate_guard: utils::sync::gate::GateGuard,
 }
 
+type BufferedWriter = owned_buffers_io::write::BufferedWriter<
+    IoBufferMut,
+    TempVirtualFileCoOwnedByEphemeralFileAndBufferedWriter,
+>;
+
+/// A TempVirtualFile that is co-owned by the [`EphemeralFile`]` and [`BufferedWriter`].
+///
+/// (Actually [`BufferedWriter`] internally is just a client to a background flush task.
+/// The co-ownership is between [`EphemeralFile`] and that flush task.)
+///
+/// Co-ownership allows us to serve reads for data that has already been flushed by the [`BufferedWriter`].
+#[derive(Debug, Clone)]
+struct TempVirtualFileCoOwnedByEphemeralFileAndBufferedWriter {
```

Comment:
This seems a bit ad hoc and redundant. Can we either add an `AsyncRead` bound to `BufferedWriter` and allow reads (are there even any callers that don't satisfy `AsyncRead`?), or do something similar in a `BufferedReadWriter` wrapper struct?

Reply:
We don't use standard tokio AsyncRead with tokio-epoll-uring, it's owned buffers style IO.

I think the idea with the wrapper struct has legs, let me try.

Reply:
Actually no:
- EphemeralFile effectively is the `BufferedReadWriter` wrapper you asked for
- and the co-ownership is necessary because there are truly two owners:
  - one owner is the writer to the VirtualFile, i.e., the flush task that is spawned by BufferedWriter
  - the other owner is whoever wants to do the reads from the flushed suffix